designs/extensible-cli-and-scaffolding-plugins-phase-1-5.md

```diff
@@ -0,0 +1,286 @@
+therefore, use this error to tell the CLI that no further execution step should be called for itself.
+
+### Initialization methods
+
+#### Update metadata
+This method will be used for two purposes. It provides CLI-related metadata to the Subcommand (e.g., 
+command name) and update the subcommands metadata such as the description or examples.
+
+- Required/optional
+  - [ ] Required
+  - [x] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+- Order guarantees
+  - [x] Method order guarantee: first method.
+  - [x] Step order guarantee:   first method.
+  - [x] Plugin order guarantee.
+
+#### Bind flags
+This method will allow subcommands to define specific flags.
+
+- Required/optional
+  - [ ] Required
+  - [x] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+- Order guarantees
+  - [x] Method order guarantee.
+  - [ ] Step order guarantee.
+  - [x] Plugin order guarantee.
+
+### Execution methods
+
+#### Inject configuration
+This method will be used to inject the `Config` object that the plugin can modify at will.
+The CLI will create/load/save this configuration object.
+
+- Required/optional
+  - [ ] Required
+  - [x] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+- Order guarantees
+  - [x] Method order guarantee.
+  - [x] Step order guarantee.
+  - [ ] Plugin order guarantee.
+
+#### Inject resource
+This method will be used to inject the `Resource` object.
+
+- Required/optional
+  - [x] Required
+  - [ ] Optional
+- Subcommands
+  - [ ] Init
+  - [ ] Edit
+  - [x] Create API
+  - [x] Create webhook
+- Order guarantees
+  - [x] Method order guarantee.
+  - [x] Step order guarantee.
+  - [ ] Plugin order guarantee.
+
+#### Pre-scaffold
+This method will be used to take actions before the main scaffolding is performed, e.g. validations.
+NOTE: a filesystem abstraction will be passed to this method that must be used for scaffolding.
+
+- Required/optional
+  - [ ] Required
+  - [x] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+- Order guarantees
+  - [x] Method order guarantee.
+  - [x] Step order guarantee.
+  - [x] Plugin order guarantee.
+
+#### Scaffold
+This method will be used to perform the main scaffolding.
+NOTE: a filesystem abstraction will be passed to this method that must be used for scaffolding.
+
+- Required/optional
+  - [x] Required
+  - [ ] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+- Order guarantees
+  - [x] Method order guarantee.
+  - [x] Step order guarantee.
+  - [x] Plugin order guarantee.
+
+#### Post-scaffold
+This method will be used to take actions after the main scaffolding is performed, e.g. cleanup.
+NOTE: a filesystem abstraction will **NOT** be passed to this method, it should interact directly.
```

Comment:
is the recommendation that you not interact with the filesystem here then, or is this just because naturally this takes place after KB writes stuff to disk, or both?

Reply:
It was because the scaffolding is expected to be done in the previous step, post-scaffold was designed to do tasks like code generation, make targets and things like that that do not require a filesystem. Even though the current design and implementation of plugin phase 1.5 do not pass a filesystem, maybe we want to leave the option to pass a filesystem in the future if the need arises.

Reply:
I updated the note saying that we may need to pass a filesystenm abstraction in a future. It is more a "We don't need it now" that a "We should not have a filesystem abstraction here".